rhombus-lib/rhombus/private/amalgam/filesystem.rhm

```diff
@@ -47,5 +71,344 @@ namespace filesystem:
   fun link_exists(p :: PathString):
     rkt.#{link-exists?}(p)
 
-  fun list_directory(p :: PathString = Path.current_directory()) :~ List.of(Path):
-    List(& rkt.#{directory-list}(p))
+  fun type(p :: PathString,
+           ~must_exist = #false):
+    ~who: who
+    match (rkt_error.local_error_adjust { #'#{file-or-directory-type}: who }:
+             rkt.#{file-or-directory-type}(p))
+    | #'#{directory-link}: #'directory_link
+    | s: s
+
+  fun files(start_p :: PathString = Path.current_directory(),
+            ~recur = #false,
+            ~follow_links = #false,
+            ~skip: skip :: Path -> Any = fun (x): #false,
+            ~add_path = #false) :~ List.of(Path):
+    ~who: who
+    fun files_here(accum :~ List, p, onto_path):
+      let files :~ PairList:
+        rkt_error.local_error_adjust { #'#{directory-list}: who }:
+          rkt.#{directory-list}(p, #{#:build?}: !onto_path && add_path)
+      for values (accum :~ List = accum) (file: files):
+        let file = (if onto_path | Path.add(onto_path, file) | file)
+        skip_when skip(file)
+        let new_accum = accum.add(file)
+        if recur && (if follow_links
+                     | directory_exists(file)
```

Comment:
If `start_p` is an absolute path more than one directory deep, and `add_path` is `#false` then `recur` has no effect.  The `file` that `directory_exists` checks will be absolute, but not attached to the current directory.

Similarly if called `files("a/b", ~recur: #true)` and there is a directory under the current directory called `b` the wrong `b` will be traversed.

Reply:
Thanks! I haven't written any tests, yet, so I'm pretty sure it mostly doesn't work. 😄

Reply:
It mostly works :D.